### ISSUE TITLE
view::join is quite slow

### ISSUE THREAD

**Issue by h-2** [ericniebler/range-v3#700]
```cpp
std::vector<std::vector<char>> two_dim;
// fill two_dim with lots of data

std::vector<char> foo = two_dim | view::join;
```
is more than three times slower than:
```cpp
std::vector<std::vector<char>> two_dim;
// fill two_dim with lots of data

std::vector<char> foo;
for (auto && s : two_dim)
     foo.insert(foo.end(), s.begin(), s.end());
```

Is this
* unexpected; *or*
* expected and unavoidable; *or*
* expected, but avoidable?

**Comment by h-2** [comment]
Ok, some more details:

I am writing a vector-of-vector replacement that has its element concatenated internally (better caching behviour, faster element access) and gives element access through a slice-view onto the underlying vector. This all works fine. The data structure is of course notoriously bad for random inserts, because it needs to copy the entire concatenation.
I still want to offer inserts, though. This is what the insert function for count times element looks like:
```cpp
    iterator insert(const_iterator pos, size_type const count, value_type const & value)
    {
        auto const pos_as_num = std::distance(cbegin(), pos); // we want to insert BEFORE this position

        if (count == 0)
            return begin() + pos_as_num;


        auto concatenated = ranges::view::repeat_n(value, count)
                          | ranges::view::join
                          | ranges::view::bounded;

        data_values.reserve(data_values.size() + count * ranges::size(value));
        data_values.insert(data_values.cbegin() + data_delimiters[pos_as_num],
                           ranges::begin(concatenated),
                           ranges::end(concatenated));

        data_delimiters.reserve(data_values.size() + count);
        data_delimiters.insert(data_delimiters.cbegin() + pos_as_num,
                               count,
                               *(data_delimiters.cbegin() + pos_as_num));

        // adapt delimiters of inserted
        for (size_type i = 0; i < count; ++i)
            data_delimiters[pos_as_num + i + 1] += ranges::size(value) * (i + 1);

        // adapt delimiters after that
        std::for_each(data_delimiters.begin() + pos_as_num + count + 1,
                      data_delimiters.end(),
                      [full_len = ranges::size(value) * count] (auto & d) { d += full_len; });

        return begin() + pos_as_num;
    }
```
This inserts n elements. The function for inserting a single element just forwards to `insert(it, 1, value)`.
This was about **100x slower than the previous implementation** (without ranges). If I copy and paste the above code for the one-element-insert, but replace
```cpp
        auto concatenated = ranges::view::repeat_n(value, count)
                          | ranges::view::join
                          | ranges::view::bounded;
```
with 
```cpp
        auto const & concatenated = value;
        size_type count = 1;
```
i.e. I remove the detour through the range (which isn't necessary), then I get a **100x speed-up** for my bench-mark.

My conclusion is that iterating and dereferencing over `ranges::view::repeat_n(rng, 1)  | ranges::view::join | ranges::view::bounded` is a hundred times slower than just iterating and dereferencing over rng 😱 

What can I do to track this down?

**Comment by ericniebler** [comment]
Which compiler/version, and with what switches? Can you post your benchmark code?

**Comment by gnzlbg** [comment]
That seems consistent with the other performance issues being reported here.

TL;DR: while most compilers can vectorize nested for loops, they cannot vectorize the state machines generated by range-v3.

What you can do: make sure you are compiling with the right flags (e.g. `-O3/fast -DNDEBUG`), make sure that you have the proper vectorizers enabled (e.g. LLVM's polly), use the vectorization diagnostics of your compiler to find out _why_ the loop isn't vectorized (clang and gcc have these), increase the inlining threshold of your compiler to see if that fixes it and fill a compiler bug...

Otherwise, give enough information to reproduce Which compiler are you using? Which compilation flags? What is the exact MWE that you are compiling to reproduce the issue? Timings, etc. so that we can check it out.

**Comment by h-2** [comment]
> Which compiler/version, and with what switches?

g++ 7.1.1 20170713 with the flags `-std=c++17 -fconcepts -O3 -DNDEBUG -Wall -Wextra`

>  Can you post your benchmark code?

It is quite a lot of code, I will try to reduce it to a more manageable size.

> That seems consistent with the other performance issues being reported here.

Hmpf, I am not sure if that's good news or bad news 🙄

**Comment by h-2** [comment]
Ok, here is an example:
```cpp
#include <iostream>
#include <chrono>
#include <random>

#include <range/v3/view/bounded.hpp>
#include <range/v3/view/const.hpp>
#include <range/v3/view/join.hpp>
#include <range/v3/view/repeat_n.hpp>
#include <range/v3/view/slice.hpp>

int main()
{
    constexpr unsigned count = 1'000;
    constexpr unsigned size = 1'000;
    using value_type = std::vector<char>;

    std::random_device rd;  //Will be used to obtain a seed for the random number engine
    std::mt19937 gen(rd()); //Standard mersenne_twister_engine seeded with rd()
    std::uniform_int_distribution<> dis(0, 3);

    // random values
    std::vector<value_type> vec;
    vec.resize(count);
    for (auto & s : vec)
    {
        s.resize(size);
        for (auto & c : s)
            c = dis(gen);
    }

    std::vector<char> target;
    target.shrink_to_fit();

    std::cout << "Insert directly: ";
    auto start = std::chrono::system_clock::now();
    for (auto && val : vec)
    {
        target.insert(target.cbegin(), ranges::begin(val), ranges::end(val));
    }
    auto end = std::chrono::system_clock::now();
    std::cout << std::chrono::duration_cast<std::chrono::microseconds>(end - start).count() << '\n';

    std::cout << target.size() << '\n';
    target.clear();
    target.shrink_to_fit();

    std::cout << "Insert from range: ";
    start = std::chrono::system_clock::now();
    for (auto && val : vec)
    {
        auto concatenated = ranges::view::repeat_n(val, 1)
                          | ranges::view::join
                          | ranges::view::bounded;
        target.insert(target.cbegin(), ranges::begin(concatenated), ranges::end(concatenated));
    }
    end = std::chrono::system_clock::now();
    std::cout << std::chrono::duration_cast<std::chrono::microseconds>(end - start).count() << '\n';

    std::cout << target.size() << '\n';

    return 0;
}

}
```

Output:
```
Insert directly:       31664
1000000
Insert from range: 32668978
1000000
```

Do you really think this can be explained by vectorization? I am not even building with `-march=native` .

**Comment by h-2** [comment]
I just tried with clang-4 (I didn't previously try because my project requires the Concepts TS), and clang is much faster:
```
Insert directly:    31688
1000000
Insert from range: 887549
1000000
```
A factor of 20 (which isn't great but very different from the factor 1000x with GCC).

**Comment by h-2** [comment]
I have reported this upstream with GCC:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=81476

But I am still worried by the 20x slow-down with Clang. Do you think this is a lower bound for how close you can get to the original performance with the current ranges design or are these further performance issues with clang?
Do you do any kind of benchmarking yourself, like basic iterations on individual view or combinations?

I am not asking, because I am expecting something, but we develop performance-focused code and something like a slow-down of factor 1-2x on iterating might be acceptable for the added quality of code and generalisation, but 20x is kinda scary.

**Comment by gnzlbg** [comment]
Can you replace the `.shrink_to_fit()` call with a `.reserve(vec.size())`? (that should isolate a bit more the impact of the ranges approach from memory allocations)

**Comment by h-2** [comment]
@gnzlbg I just tried that a couple of minutes ago, the allocations are entirely irrelevant for the numbers. Also, interestingly, the behaviour scales (or "unscales") to lower optimisation levels, as well, as I posted in the GCC issue.

**Comment by gnzlbg** [comment]
I think your bug report is "too complicated" for compiler writers, as in, they need to get range-v3, the generated code probably contains 1 million lines of code, etc.

The next steps here (or the tensor indices bug, which might be easier) are something like:

- desugaring the range-v3 code into a self-contained state machine (e.g. two concrete iterator types with as minimum code as possible to reproduce this, no dependencies). Verify that range-v3 doesn't have a performance bug in this step.
- write the equivalent nested for-loop type of code
- compare the generated code, IR, assembly, ... between both after applying the optimizations. These perf problems look to me like the compiler doesn't recognize the iterator-like code as a sequence of nested raw-loops and then gives up.
- fill bugs in the compilers in which the performance doesn't match (probably all of them...) and hope that they are fixable.  While vectorizing nested loops is "easy", I don't know how hard it is to transform a sequence of `while (!...) { while(!...) { .... } ... }` into the canonical loop form of the compiler. While it might be possible, the solution might turn out to be "too hard".

**Comment by ericniebler** [comment]
I disagree that the report is too complicated for a compiler dev to look at. All they would need is a preprocessed source file, the compiler switches, and the compiler version.

`view::join` is expensive. There's no getting around it. In order to make N views look like 1, you need to test at every increment whether you have reached the end of one range and need to skip to the next.

In addition, the join view reuses memory by storing things in a variant. That probably kills alias analysis. 

The quick fix for this is for performance sensitive code to avoid joins. The long term fix would be to add support for segmented ranges and hierarchical algorithms.

**Comment by h-2** [comment]
> I disagree that the report is too complicated for a compiler dev to look at. All they would need is a preprocessed source file, the compiler switches, and the compiler version.

The GCC people were indeed able to figure it out, the dramatic slowdown over clang seems to actually be a bug in libstdc++'s vector's insert (see the link above). [so its a libc++ vs libstdc++ issue, not clang vs gcc]

> `view::join` is expensive. There's no getting around it. In order to make N views look like 1, you need to test at every increment whether you have reached the end of one range and need to skip to the next.
>
> In addition, the join view reuses memory by storing things in a variant. That probably kills alias analysis.
>
> The quick fix for this is for performance sensitive code to avoid joins. The long term fix would be to add support for segmented ranges and hierarchical algorithms.

Ok, thanks for the details! The last bit sounds quite interesting.

I will probably avoid the join view for now. Since I only have sized random access ranges as input, I might also be able to hack a less generic version of join together that performs better for my use cases...let's see.

**Comment by ericniebler** [comment]
> Since I only have sized random access ranges as input, I might also be able to hack a less generic version of join together that performs better for my use cases...let's see.

I'm curious what you have in mind. If you make headway on that, please share back. A generic solution doesn't preclude an optimization for random access ranges.

**Comment by gnzlbg** [comment]
> The GCC people were indeed able to figure it out

The slowdown of gcc with respect to clang, or the 20x slowdown wrt not
using join?

**Comment by ericniebler** [comment]
My hunch is that once the libstdc++ bug is fixed, gcc will outperform clang. See the bottom of [this blog post](http://ericniebler.com/2014/04/27/range-comprehensions/), for instance.

You may also run into the gcc optimizer bug discussed [here](https://twitter.com/kwalfridsson/status/877260401775448065).

**Comment by h-2** [comment]
> The slowdown of gcc with respect to clang, or the 20x slowdown wrt not using join?

> My hunch is that once the libstdc++ bug is fixed, gcc will outperform clang. See the bottom of this blog post, for instance.

Yes, their implemented fix reduces the slowdown to 1.2x wrt to not using join according to the GCC bug tracker which is very good. 

One more thing: the GCC devs said the issue was only triggered because the iterators returned were pure input_iterators and not even forward_iterators... is there a reason they can't be forward or even bi-directional? To be honest, which range properties are preserved, and which are lost, is one of the things I currently find most difficult to quickly assess, but I will open a separate issue for that.

**Comment by jwakely** [comment]
The example is pretty dumb, it would be a lot better to insert at the end of `target` and not keep pushing into the beginning. That's a very poor algorithm.

````
    std::cout << "Insert from range (rev): ";
    start = std::chrono::system_clock::now();
    for (auto && val : ranges::view::reverse(vec))
    {
        auto concatenated = ranges::view::repeat_n(val, 1)
                          | ranges::view::join
                          | ranges::view::bounded;
        target.insert(target.cend(), ranges::begin(concatenated), ranges::end(concatenated));
    }
    end = std::chrono::system_clock::now();
    std::cout << std::chrono::duration_cast<std::chrono::microseconds>(end - start).count() << '\n';

    std::cout << target.size() << '\n';
    target.clear();
    target.shrink_to_fit();
````

This is a huge improvement, using any implementation of `std::vector`.

With libstdc++:

Insert directly: 12620
1000000
Insert from range (rev): 2265
1000000
Insert from range: 11446412
1000000

With libc++:

Insert directly: 13584
1000000
Insert from range (rev): 3843
1000000
Insert from range: 676498
1000000

**Comment by ericniebler** [comment]
> Yes, their implemented fix reduces the slowdown to 1.2x wrt to not using join according to the GCC bug tracker which is very good.

Wow! 👍👍👍👍

**Comment by h-2** [comment]
> I'm curious what you have in mind. If you make headway on that, please share back. A generic solution doesn't preclude an optimization for random access ranges.

@ericniebler I have implemented a join view that preserves random access and sized'ness. It is slower than the your implementation (can't beat factor 1.2x), but the asymptotic run-times are okay and I am still optimising things.
I have implemented this in our library, because it relies on third-party data structures for core functionality. Descriptions, asymptotic complexity and bench-marks can be found in seqan/seqan3#104

**Comment by CaseyCarter** [comment]
> In addition, the join view reuses memory by storing things in a variant. That probably kills alias analysis.

Only the "join and inject delimiters" specialization uses a variant, so that fact isn't germane to these numbers. That said, using a variant to reduce the total size of the join view iterators was likely a poor choice, it would be an improvement to move all of that state from the iterators into the view itself.

**Comment by ericniebler** [comment]
> That said, using a variant to reduce the total size of the join view iterators was likely a poor choice, it would be an improvement to move all of that state from the iterators into the view itself.

IIRC, the join view originally kept all its state in the view. I moved some state to the iterators because the Pythagorean triples example was faster that way.

**Comment by CaseyCarter** [comment]
> I moved some state to the iterators because the Pythagorean triples example was faster that way.

My recollection is that the example was blazingly fast with the inner range stored in the iterator, which violated all sorts of semantic requirements. You had changed it to be stored in the outer range before I did the rework that used `variant` for the "join and inject delimiters" case.
